src/charm.py

```diff
@@ -0,0 +1,469 @@
+
+        app_type = self.config.app_type
+        logger.info(f"app: {app_type} : {type(app_type)}")
+        username = event.params["username"]
+        password = event.params["password"]
+        servers = event.params["servers"]
+        topic = event.params["topic_name"]
+        consumer_group_prefix = None
+        if "consumer_group_prefix" in event.params:
+            consumer_group_prefix = event.params["consumer_group_prefix"]
+
+        pid = self._start_process(
+            process_type=app_type,
+            username=username,
+            password=password,
+            servers=servers,
+            topic=topic,
+            consumer_group_prefix=consumer_group_prefix,
+            tls="disabled",
+        )
+        event.set_results({"process": pid})
+
+    def _stop_process_action(self, event: ActionEvent):
+        """Handle the stop process action."""
+        if not self.peer_relation.unit_data.pid:
+            event.fail("No process to stop!")
+            return
+
+        pid = self.peer_relation.unit_data.pid
+        r_pid = self._stop_process(pid)
+        assert r_pid == pid
+        event.set_results({"process": pid})
+
+    def _start_process(
+        self,
+        process_type: AppType,
+        username: str,
+        password: str,
+        servers: str,
+        topic: str,
+        consumer_group_prefix: Optional[str],
+        tls: str,
+        cafile_path: Optional[str] = None,
+        certfile_path: Optional[str] = None,
+        keyfile_path: Optional[str] = None,
+    ) -> int:
+        """Handle the start of the process for consumption or production of messagges."""
+        t0 = int(time())
+        my_cmd = f"{self._build_cmd(process_type, username, password, servers, topic, consumer_group_prefix, tls, cafile_path, certfile_path, keyfile_path)}"
+        logger.info(my_cmd)
+        process = subprocess.Popen(
+            shlex.split(my_cmd),
+            stdout=open(f"/tmp/{t0}_{process_type.value}.log", "w"),
+            stderr=open(f"/tmp/{t0}_{process_type.value}.err", "w"),
+        )
+        self.peer_relation.set_pid(process.pid)
+        return process.pid
+
+    def _stop_process(self, pid: int):
+        """Handle the stop of the producer/consumer process."""
+        logger.info(f"Killing process with pid: {pid}")
+        subprocess.Popen(["sudo", "kill", "-9", str(pid)])
+        self.peer_relation.remove_pid(pid)
+        return pid
+
+    def _build_cmd(
+        self,
+        process_type: AppType,
+        username: str,
+        password: str,
+        servers: str,
+        topic: str,
+        consumer_group_prefix: Optional[str],
+        tls: str,
+        cafile_path: Optional[str] = None,
+        certfile_path: Optional[str] = None,
+        keyfile_path: Optional[str] = None,
+    ):
+        """Handle the creation of the command to launch producer or consumer."""
+        if tls == "enabled" and self.peer_relation.app_data.private_key:
```

Comment:
**issue/todo/thought** Why not having a flag "self.tls_enabled" which is auto-generated based on whether a relation exists? Wouldn't this make things "safer", self-driven and even easier from an implementation standpoint?

Reply:
Yes, I guess it realted to the previous comment. Let's decide what we want to do and then approach these accordintly